### ISSUE TITLE
function that takes generic with no constraint can call function with type constraint with invalid type

### ISSUE THREAD

**Issue by DetachHead** [microsoft/TypeScript#43956]
# Bug Report

### 🔎 Search Terms
Generic no constraint call function with no constraint

### 🕗 Version & Regression Information
4.3.0-dev-20210504

### ⏯ Playground Link

https://www.typescriptlang.org/play?ts=4.3.0-dev.20210506#code/GYVwdgxgLglg9mABAIwIYCcA8AVRBTADyjzABMBnRAbwG0BrPATwC5Fyp0YwBzAXVdRhGAXwB8ACgBuqADYg8rbAEpqAKEQbE0uXgB0AC1TkA8gHcwABXRwADnnRRG4gOTA4cZ0tXDVq0JFgERDc4HAlteUUlKnVNNCxscNl5FQB6VLA4fHRrdAAafEkSRCh9OBBufURcUjg8cjBnKEQAW1QoCCrSvEQIBHZ0VC4ob1948XBSPGAuPFI01L6cvGgZRmzc8lZJ6dnSRBhKTOajchhuMFRkGR6oLOw-dwmyXbA5hcyNuHzgoZlKdqIdDgWAtPBAA

### 💻 Code

```ts
function bar<T extends {[key: string]: any}>(value: T) {
    value.hasOwnProperty('foo')
}

function foo<T>(value: T){
    bar<T>(value) //no error, even though T doesn't match the constraint
}

bar(undefined) //correctly errors: undefined is not assignable to T
foo(undefined) //no error, fails at runtime
```

### 🙁 Actual behavior
No compile error when calling `bar` with a type that doesn't match the bounds of `T`

### 🙂 Expected behavior
Compile error

**Comment by MartinJohns** [comment]
This sounds like #31006.

**Comment by RyanCavanaugh** [comment]
This would be a pretty noisy breaking change for likely not much upside, but we can discuss

**Comment by DetachHead** [comment]
this seems to be fixed as of 4.8

```ts
function bar<T extends {[key: string]: any}>(value: T) {
    value.hasOwnProperty('foo')
}

function foo<T>(value: T){
    bar<T>(value) //error in 4.8
}
```
